core/src/main/java/com/linecorp/armeria/common/HttpHeadersBase.java

```diff
@@ -204,8 +204,14 @@ URI uri() {
             final String scheme = scheme();
             checkState(scheme != null, ":scheme header does not exist.");
             final String authority = authority();
-            checkState(authority != null, ":authority header does not exist.");
-            uri = scheme + "://" + authority + path;
+
+            final StringBuilder sb = new StringBuilder(scheme + ":");
+            if (authority != null) {
+                sb.append("//");
+                sb.append(authority);
+            }
+            sb.append(path);
```

Comment:
- Maybe pre-calculate the length to avoid expansion?
  ```suggestion
              final StringBuilder sb = new StringBuilder(
                      scheme.length() + 1 +
                      (authority != null ? (authority.length() + 2) : 0) +
                      path.length());
              sb.append(scheme);
              sb.append(':');
              if (authority != null) {
                  sb.append("//");
                  sb.append(authority);
              }
              sb.append(path);
  ```
- If `authority` is `null`, `uri` could be a string like `http:/path`. Is such a string a valid URI? If not (i.e. `new URI()` below fails), shouldn't we raise an exception before we construct a string?

Reply:
Ah, I see. No worries, then.